kong/plugins/acl/groups.lua

```diff
@@ -138,9 +124,56 @@ local function get_current_consumer_id()
 end
 
 
+--- Returns a table with all group names.
+-- The table will have an array part to iterate over, and a hash part
+-- where each group name is indexed by itself. Eg.
+-- {
+--   [1] = "users",
+--   [2] = "admins",
+--   users = "users",
+--   admins = "admins",
+-- }
+-- If there are no groups defined, it will return an empty table
+-- @param ctx current ngx.ctx passed as an argument
+-- @return table with groups or nil
+local function get_authenticated_groups()
+  local authenticated_groups = kong.ctx.shared.authenticated_groups
+  if type(authenticated_groups) ~= "table" then
+    authenticated_groups = ngx.ctx.authenticated_groups
+    if type(authenticated_groups) ~= "table" then
+      return nil
+    end
+  end
+
+  local groups = {}
+  for i = 1, #authenticated_groups do
+    groups[i] = authenticated_groups[i]
+    groups[authenticated_groups[i]] = authenticated_groups[i]
+  end
+
+  return groups
+end
```

Comment:
The biggest thing here is `ngx.ctx` access which can be expensive. If there is no table, we don't create a new table. If there is a table, there is one already. But caching the resulting table is very premature.

Reply:
Yes, caching that is not possible since context table is always different on each request vs. dao cached entity table (which is same). That's why we don't cache. If I were to optimize this, I would remove whole lookup table creation and just loop over context table directly. Maybe I change to that even If that makes code very different for context value vs. cached dao row.